File: rust-java-bridge/swim_client/swim_client_core/src/downlink/decoder.rs
// Copyright 2015-2024 Swim Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::io::ErrorKind;

use bytes::{Bytes, BytesMut};
use swim_api::error::{FrameIoError, InvalidFrame};
use swim_api::protocol::downlink::{DownlinkNotification, DownlinkNotificationDecoder};
use swim_api::protocol::map::{extract_header, MapMessage};
use swim_recon::parser::MessageExtractError;
use tokio_util::codec::Decoder;

pub struct MapDlNotDecoder {
    delegate: DownlinkNotificationDecoder<MapMessage<Bytes, Bytes>, MapMessageDecoder>,
}

impl Default for MapDlNotDecoder {
    fn default() -> Self {
        MapDlNotDecoder {
            delegate: DownlinkNotificationDecoder::new(MapMessageDecoder),
        }
    }
}

impl Decoder for MapDlNotDecoder {
    type Item = DownlinkNotification<MapMessage<Bytes, Bytes>>;
    type Error = FrameIoError;

    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        self.delegate.decode(src)
    }
}

struct MapMessageDecoder;

impl Decoder for MapMessageDecoder {
    type Item = MapMessage<Bytes, Bytes>;
    type Error = FrameIoError;

    fn decode(&mut self, _src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        Ok(None)
    }

    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        let src = buf.clone().freeze();
        buf.clear();
        Ok(Some(extract_header(&src).map_err(|e| match e {
            MessageExtractError::BadUtf8(_) => FrameIoError::Io(ErrorKind::InvalidData.into()),
            MessageExtractError::ParseError(e) => {
                FrameIoError::BadFrame(InvalidFrame::InvalidHeader { problem: e.into() })
            }
        })?))
    }
}

pub struct ValueDlNotDecoder {
    delegate: DownlinkNotificationDecoder<Bytes, ValueMessageDecoder>,
}

impl Default for ValueDlNotDecoder {
    fn default() -> Self {
        ValueDlNotDecoder {
            delegate: DownlinkNotificationDecoder::new(ValueMessageDecoder),
        }
    }
}

impl Decoder for ValueDlNotDecoder {
    type Item = DownlinkNotification<Bytes>;
    type Error = FrameIoError;

    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        self.delegate.decode(src)
    }
}

pub struct ValueMessageDecoder;

impl Decoder for ValueMessageDecoder {
    type Item = Bytes;
    type Error = FrameIoError;

    fn decode(&mut self, _src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        Ok(None)
    }

    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        let src = buf.clone().freeze();
        buf.clear();
        Ok(Some(src))
    }
}

File: swim-java/swim-core/swim-codec/src/main/java/ai/swim/codec/decoder/Decoder.java
/*
 * Copyright 2015-2024 Swim Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ai.swim.codec.decoder;

import ai.swim.codec.data.ReadBuffer;

/**
 * An abstract, incremental, decoder that yields objects of type {@code T}.
 *
 * @param <T> the decoder's target type.
 */
public abstract class Decoder<T> {

  /**
   * Constructs a new decoder that is in a {@code done} state.
   *
   * @param decoder that decoded {@code T}. Calling {@link Decoder#reset} on the done decoder will return this.
   * @param value   that the decoder produced.
   * @param <T>     the decoder's target type.
   * @return a decoder in the done state.
   */
  public static <T> Decoder<T> done(Decoder<T> decoder, T value) {
    return new DecoderDone<>(decoder, value);
  }

  protected static int longToInt(long len) throws DecoderException {
    try {
      return Math.toIntExact(len);
    } catch (ArithmeticException e) {
      // Java array capacity is an integer but it is possible that Rust/a peer sends a long array.
      throw new DecoderException(e);
    }
  }

  /**
   * Returns whether the decoder is in a {@code done} state.
   */
  public boolean isDone() {
    return false;
  }

  /**
   * Feed this decoder some data to decode.
   *
   * @param buffer to decode.
   * @return a decoder representing the output of the operation.
   * @throws DecoderException if it was not possible to decode the buffer.
   */
  public abstract Decoder<T> decode(ReadBuffer buffer) throws DecoderException;

  /**
   * If this decoder is in a {@code done} state, take the decoded value.
   *
   * @return the decoded value.
   * @throws IllegalStateException if this decoder is not in a done state.
   */
  public T bind() {
    throw new IllegalStateException("Decoder is not in a done state");
  }

  /**
   * Reset this decoder back to its original state.
   */
  public abstract Decoder<T> reset();

}
